content/articles/How-to-start-up-with-Socket-Programming-with-Java/index.md

```diff
@@ -0,0 +1,331 @@
+```
+
+localhost - is a domain name like google.com, localhost takes you to your computer. It is resolved to IP address 127.0.0.1 .
+port - Port – This is built to connect with the server via the port number socket object.
+Two computers communicate through sockets.
+
+Client and Server communicate by writing to an output stream and reading from an input stream on the socket.
+
+```java
+inputStreamReader = new InputStreamReader(socket.getInputStream());  
+outputStreamWriter = new OutputStreamWriter(socket.getOutputStream());
+``` 
+
+The Scanner class is used to get input from the user from the console/terminal. System. in is a keyboard input connected input stream system.
+
+```java
+Scanner scanner = new Scanner(System.in);
+```
+
+### Methods used in Client Socket
+1. `public InputStream( getInputStream())` - returns an InputStream object for receiving data. Our Server will keep on receiving messages until client sends "BYE".
+2. `public OutputStream( getOutputStream())`- returns an OutputStream for sending data.
+3. `BufferReader( )` - Wraps inputStreamReader to improve efficiency
+4. `BufferWriter( )` - Wraps outputStreamReader to improve efficiency
+5. `BufferedWriter.flush( )` - Flushes the streams...The flush method is invoked when the buffer is complete.
+6.  `void close() `- closes all streams and the socket.
+  
+### THIS IS A CLIENT PROGRAM THAT WILL BE COMMUNICATING WITH THE SERVER.
+```java
+package com.company;  
+  
+import java.io.*;  
+import java.net.Socket;  
+import java.util.Scanner;  
+  
+public class client {  
+ public static void main(String[] args) {  
+
+  Socket socket = null;  
+  InputStreamReader inputStreamReader = null;  
+  OutputStreamWriter outputStreamWriter = null;  
+  BufferedReader bufferedReader = null;  
+  BufferedWriter bufferedWriter = null;  
+  
+try {  
+socket = new Socket("localhost", 5000);  
+inputStreamReader = new InputStreamReader(socket.getInputStream());  
+outputStreamWriter = new OutputStreamWriter(socket.getOutputStream());  
+
+bufferedReader = new BufferedReader(inputStreamReader);  
+bufferedWriter = new BufferedWriter(outputStreamWriter);  
+
+Scanner scanner = new Scanner(System.in);  
+
+while (true)
+ {  
+  String msgToSend = scanner.nextLine();  
+
+  bufferedWriter.write(msgToSend);  
+
+  bufferedWriter.newLine();  
+
+  bufferedWriter.flush();  
+  
+  System.out.println("Server: " + bufferedReader.readLine());  
+  
+if (msgToSend.equalsIgnoreCase("BYE"))  
+    break;  
+  }  
+} catch (IOException e) {  
+   e.printStackTrace();  
+} finally {  
+try {  
+
+if (socket != null)  
+ socket.close();  
+if (inputStreamReader != null)  
+  inputStreamReader.close();  
+if (outputStreamWriter != null)  
+ outputStreamWriter.close();  
+if (bufferedReader != null)  
+ bufferedReader.close();  
+if (bufferedWriter != null)  
+ bufferedWriter.close();  
+} catch (IOException e) {  
+  e.printStackTrace();  
+          }  
+        }  
+    }  
+}
+```
+The output of the code:
```

Comment:
Consider rewording this to "A sample output of this code", since the code could have different outputs depending on what the client types.